test/parallel_api/iterator/device_copyable.pass.cpp

```diff
@@ -0,0 +1,627 @@
+    static_assert(!std::is_trivially_copy_constructible_v<int_device_copyable>,
+                  "int_device_copyable is not trivially copy constructible");
+    static_assert(!std::is_trivially_copy_constructible_v<noop_device_copyable>,
+                  "noop_device_copyable is not trivially copy constructible");
+    static_assert(!std::is_trivially_copy_constructible_v<constant_iterator_device_copyable>,
+                  "constant_iterator_device_copyable is not trivially copy constructible");
+
+    static_assert(sycl::is_device_copyable_v<int_device_copyable>, "int_device_copyable is not device copyable");
+    static_assert(sycl::is_device_copyable_v<noop_device_copyable>, "noop_device_copyable is not device copyable");
+    static_assert(sycl::is_device_copyable_v<constant_iterator_device_copyable>,
+                  "constant_iterator_device_copyable is not device copyable");
+
+    //now check that our custom types can be device copyable with these test types
+    //transform_iterator
+    static_assert(sycl::is_device_copyable_v<
+                      oneapi::dpl::transform_iterator<constant_iterator_device_copyable, noop_device_copyable>>,
+                  "transform_iterator is not device copyable with device copyable types");
+    //zip_iterator
+    static_assert(sycl::is_device_copyable_v<oneapi::dpl::zip_iterator<constant_iterator_device_copyable, int*>>,
+                  "zip_iterator is not device copyable with device copyable types");
+    //permutation_iterator
+    static_assert(
+        sycl::is_device_copyable_v<
+            oneapi::dpl::permutation_iterator<constant_iterator_device_copyable, constant_iterator_device_copyable>>,
+        "permutation_iterator is not device copyable with device copyable types");
+    static_assert(sycl::is_device_copyable_v<
+                      oneapi::dpl::permutation_iterator<constant_iterator_device_copyable, noop_device_copyable>>,
+                  "permutation_iterator is not device copyable with device copyable types");
+    //custom_brick
+    static_assert(sycl::is_device_copyable_v<oneapi::dpl::internal::custom_brick<
+                      noop_device_copyable, int_device_copyable, oneapi::dpl::internal::search_algorithm::lower_bound>>,
+                  "custom_brick is not device copyable with device copyable types");
+    //replace_if_fun
+    static_assert(
+        sycl::is_device_copyable_v<oneapi::dpl::internal::replace_if_fun<int_device_copyable, noop_device_copyable>>,
+        "replace_if_fun is not device copyable with device copyable types");
+    //scan_by_key_fun
+    static_assert(
+        sycl::is_device_copyable_v<
+            oneapi::dpl::internal::scan_by_key_fun<int_device_copyable, int_device_copyable, noop_device_copyable>>,
+        "scan_by_key_fun is not device copyable with device copyable types");
+    //segmented_scan_fun
+    static_assert(sycl::is_device_copyable_v<oneapi::dpl::internal::segmented_scan_fun<
```

Comment:
I am not sure how else we could do the same, but do we really want to use our internals in this test?

Reply:
I thought about this when adding these, but I don't have a better way of doing this.  I know we have had a general decision against doing this.  However, I think this situation deserves targeted "unit testing" and that requires reaching into the internals in this case.  

We could try to do it all via setting up contrived examples with system level tests, but that would take a lot more thought and effort for the same coverage, and result in a more computationally wasteful test with more redundancy with the rest of the test suite.  

I would be interested in establishing some "unit test" namespace, or suite of tests, where reaching into the internals was not frowned upon.  It can be quite useful for some situations like this.

Reply:
Basically, I agree that "using" internals in a system level test should not be done.  
However, "testing" internals in a unit level test should be a tool we are allowed to use toward the goal of well tested software.

Reply:
The problem is, we do not clearly separate and not even have clear definitions of "system level" vs "unit level" tests. Perhaps I would prefer to name it "API conformance" and "implementation detail" tests.

If it's an "implementation detail" test, I prefer this to be clearly marked as such in the name, or even better put to a separate directory.

Reply:
Sure, we can make that change.  Do you think that we should then separate into two files the testing of the custom iterator types from the testing of internal helper types?
Or should we move all testing of device copyable trait into an "implementation_detail" folder?

Reply:
Testing the oneDPL iterator types for copyability makes most sense in their respective API tests - essentially, it's part of their expected semantics (which may need to be better documented/specified, but anyway).

Reply:
Ah, that is a good plan.  I'll need to move the helpers to some central location so they can be used in both tests.

Reply:
This is done.  I moved the tests to a folder called "implementation_details", and moved the custom iterator tests to their respective public API tests.  
I also replaced the old check for device copyable or legacy device copyable with this new static assert., and removed the utility as we should not need or rely on it anymore.